### ISSUE TITLE
Proposal: Add JSX-specific overrides to `no-extra-parens`

### ISSUE THREAD

**Issue by Alaneor** [eslint/eslint#7444]
In order to accommodate best practices established in the JSX world, the rule `no-extra-parens` needs to be modified to optionally ignore `JSX` components - the general standard is to wrap at least multiline JSX components in parens. However, this conflicts with the best practices in JavaScript in general where extra parens around expressions are discouraged.

**What rule do you want to change?**

[`no-extra-parens`](http://eslint.org/docs/rules/no-extra-parens)

**Does this change cause the rule to produce more or fewer warnings?**

Neither - the default setting should not cause more or fewer warnings to appear. However, this proposal will allow users to reduce some warnings when working with JSX components.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option: `ignoreJSX` with possible values:
- `all`: All JSX components are **excluded** from this rule
- `multiline`: All multiline JSX components are **excluded** from this rule
- `single-line`: All single-line JSX components are **excluded** from this rule

The default behaviour for this exception will be to to disallow extra parens around any `JSX` components, just as it is currently implemented.

**Please provide some example code that this change will affect:**

``` js
/* eslint no-extra-parens: ["error", "all", { "ignoreJSX": "all" }] */
// Ok - JSX components are excluded from this rule
const Component = (<div />)
const Component = <div />
const Component = (
  <div>
    <p />
  </div>
)
const Component = 
  <div>
    <p />
  </div>

/* eslint no-extra-parens: ["error", "all", { "ignoreJSX": "multiline" }] */
// Ok (extra parens not allowed)
const Component = <div />
// Ok (multiline JSX components are excluded from this rule)
const Component = (
  <div>
    <p />
  </div>
)
const Component = 
  <div>
    <p />
  </div>
// Warns (extra parens not allowed)
const Component = (<div />)

/* eslint no-extra-parens: ["error", "all", { "ignoreJSX": "single-line" }] */
// Ok (single-line JSX components are excluded from this rule)
const Component = <div />
const Component = (<div />)
// Ok (extra parens not allowed)
const Component = 
  <div>
    <p />
  </div>
// Warns (extra parens not allowed)
const Component = (
  <div>
    <p />
  </div>
)
```

**What does the rule currently do for this code?**

The rule currently flags all cases where a JSX component is wrapped in parens. The general best practice, however, is to wrap at least multiline JSX components in parens, for better readability.

**What will the rule do after it's changed?**

The rule will allow developers to **optionally** turn this rule off for `JSX` components (single-line, multiline or all). The default setting will be to treat parens around `JSX` components just like any other expression - it will warn if `JSX` components are wrapped in parens.

See also original discussion in #7350.

_/cc @platinumazure @ljharb_

**Comment by ljharb** [comment]
This seems to preclude setting `jsx` to both single-line and multiline? I feel like it might be clearer to have three separate boolean options - `ignoreJSX`, `ignoreSingleLineJSX`, and `ignoreMultilineJSX` - where of course `ignoreJSX` being `true` would make the other two settings irrelevant.

**Comment by Alaneor** [comment]
@ljharb You can set it to `false` to ignore **all** jsx components. Is that what you are looking for?

**Comment by ljharb** [comment]
ahhh yes i see now, thanks for clarifying. Perhaps instead of letting the "jsx" be `false`, we have it be an enum of "none", "all", "single-line", and "multiline"?

**Comment by platinumazure** [comment]
Gah my email didn't go through. I had proposed this:

"ignoreJSX": "all" | "single-line" | "multi-line"

I don't like "jsx" as an option name because it's not clear if the option would increase or decrease lint warnings around JSX.

**Comment by Alaneor** [comment]
@ljharb The `false` option was chosen to match the other overrides available for this rule (check the docs linked above). This rule will flag extra parens, and you can "disable" this rule for some parts of the code, thus `false`. The opposite, `true` or `"all"` thus does not fit well in this rule because that is the default behaviour anyway.

**Comment by platinumazure** [comment]
@Alaneor Boolean makes sense when that's literally the entire scope of the option-- ignore all of something, or ignore none. (Or warn all, warn none.) The JSX option is clearly not boolean. That's why I recommend a string enum option.

**Comment by Alaneor** [comment]
@platinumazure Okay, makes sense. I still feel the option to disable parens around all JSX components should be `"none"`, though - it should read so as to "remove extra parens around **no** JSX components", not all of them.

**Comment by platinumazure** [comment]
@Alaneor The correct option value depends on the option name, which is why I made a point of including a verb in the option. For "ignoreJSX", "all" clearly means don't lint any JSX components (ignore all JSX components). If the option was "lintJSX" or "enforceOnJSX" etc., then "none" makes sense.

**Comment by platinumazure** [comment]
We can bikeshed on the option name. My concern with the current proposed
name is the lack of a verb (are we ignoring/tolerating JSX or enforcing it
more strictly?). This also means false is a boolean trap.

My proposal:

"ignoreJSX": "all" | "single-line" | "multi-line"

Thoughts?

**Comment by Alaneor** [comment]
Let's not do any more bikeshedding - I have updated the proposal as per @platinumazure's suggestion.

**Comment by Alaneor** [comment]
Is there anything I can do to move this proposal forward? Thanks!

**Comment by kaicataldo** [comment]
@Alaneor Sorry, seems like we lost track of this. I've put my thumb on this. @eslint/eslint-team Any other thoughts?

**Comment by vitorbal** [comment]
It seems like all we need now is a champion.

**Comment by kaicataldo** [comment]
I'll champion this

**Comment by kaicataldo** [comment]
@Alaneor Sorry again for the delay, but this issue is now accepted and is ready for implementation! Are you willing and able to implement this?

**Comment by Alaneor** [comment]
Yes, already working on this! 🎉 I actually have it ready now, only documentation remains to be updated. I will open a PR later today.

**Comment by Alaneor** [comment]
PR is ready for review! #7926
